### ISSUE TITLE
Subscription does not work

### ISSUE THREAD

**Issue by StefanCiobanu1989** [php-mqtt/client-examples#4]
Hello,
I'm trying to subscribe to a mosquitto broker using the given examples but it cant connect, it's either a 500 error or a timeout. Note publishing works without issues. I have used https://github.com/php-mqtt/client-examples/blob/master/02_subscribing/03_subscribe_with_qos_2.php but failed so i was wondering if this a case of outdated code. Any help would be great. Thanks !

**Comment by Namoshek** [comment]
You get a `500` error? Where? That doesn't sound very much like an MQTT issue.

The code you linked is up to date and works, I verified just to be sure (see my steps below). Can you try a different client id? And maybe also increase the log level, the output of it could help. It makes a difference whether the script fails while connecting or when subscribing.

---

To run mosquitto as broker, run from the `php-mqtt/client` repository:
```sh
$ ~/client> sh create-certificates.sh
$ ~/client> docker run --rm -it -p 1883:1883 -p 8883:8883 -p 8884:8884 -v (pwd)/.ci/tls:/mosquitto-certs -v (pwd)/.ci/mosquitto.conf:/mosquitto/config/mosquitto.conf eclipse-mosquitto:1.6.12
```

In a second tab, run the subscribe script:
```sh
$ ~/client-examples> php 02_subscribing/03_subscribe_with_qos_2.php
[2021-10-25T16:40:33+02:00] MQTT [127.0.0.1:1883] [test-subscriber] Connection with broker established successfully.
[2021-10-25T16:40:47+02:00] We received a message on topic [foo/bar/baz]: Hello world!
```

In a third tab, run the publish script (any QoS):
```sh
$ ~/client-examples> php 01_publishing/03_publish_with_qos_2.php
[2021-10-25T16:41:04+02:00] MQTT [127.0.0.1:1883] [test-publisher] Connection with broker established successfully.
```

**Comment by StefanCiobanu1989** [comment]
i changed the client id to null and different others, and it cant get anything back from the broker regarding subscriptions, the connection is ok, it works with no errors. The code i am using is :

```
<?php


require __DIR__ . '/vendor/autoload.php';
//require __DIR__ . '/../shared/config.php';

use PhpMqtt\Client\Exceptions\MqttClientException;
use PhpMqtt\Client\MqttClient;

$logger = new SimpleLogger(LogLevel::INFO);
try {
    // Create a new instance of an MQTT client and configure it to use the shared broker host and port.
    $client = new MqttClient("serverIP",
                             "1883",
                             "test-subscriber",
                             MqttClient::MQTT_3_1,
                             null);


    // Connect to the broker without specific connection settings but with a clean session.
    $client->connect(null, true);

    // Subscribe to the topic 'THETOPIC' using QoS 2.
     $client->subscribe('THETOPIC', function (string $topic, string $message, bool $retained) use ($logger, $client) {
        $logger->info('We received a {typeOfMessage} on topic [{topic}]: {message}', [
            'topic' => $topic,
            'message' => $message,
            'typeOfMessage' => $retained ? 'retained message' : 'message',
        ]);

        // After receiving the first message on the subscribed topic, we want the client to stop listening for messages.
        $client->interrupt();
    }, MqttClient::QOS_EXACTLY_ONCE);

    // Since subscribing requires to wait for messages, we need to start the client loop which takes care of receiving,
    // parsing and delivering messages to the registered callbacks. The loop will run indefinitely, until a message
    // is received, which will interrupt the loop.
    $client->loop(true);

    // Gracefully terminate the connection to the broker.
    $client->disconnect();
} catch (MqttClientException $e) {
    // MqttClientException is the base exception of all exceptions in the library. Catching it will catch all MQTT related exceptions.
    //$logger->error('Subscribing to a topic using QoS 2 failed. An exception occurred.', ['exception' => $e]);
}

```

if  "$client->loop(true);" was to be removed then the 500 error goes away but still no message come back from the broker. Upon switching $logger = new SimpleLogger(LogLevel::INFO); to $logger = new SimpleLogger(LogLevel::DEBUG); we get a message back that the connection was successful however nothing regarding the subscription part. Do i need the "require __DIR__ . '/../shared/config.php';" ?

**Comment by Namoshek** [comment]
Can you please provide the log output with `LogLevel::DEBUG`? From your description, I'm not sure where the problem could be coming from. It is not the include of `config.php` though, it is optional and only meant for central configuration of all examples at once.

**Comment by StefanCiobanu1989** [comment]
```
MQTT [ip:port] [cc432d1b33936e5c206f] Connecting to broker.
MQTT [ip:port] [cc432d1b33936e5c206f] Socket opened and ready to use.
MQTT [ip:port] [cc432d1b33936e5c206f] Using the [clean session] flag for the connection.
MQTT [ip:port] [cc432d1b33936e5c206f] Sending connection handshake to broker.
MQTT [ip:port] [cc432d1b33936e5c206f] Sent data over the socket: "MQIsdp cc432d1b33936e5c206f MQTT [ip:port] [cc432d1b33936e5c206f] Waiting for 1 bytes of data.
MQTT [ip:port] [cc432d1b33936e5c206f] Read data from the socket:
MQTT [ip:port] [cc432d1b33936e5c206f] Read data from the socket (without blocking):  
MQTT [ip:port] [cc432d1b33936e5c206f] Connection with broker established successfully.
MQTT [ip:port] [cc432d1b33936e5c206f] Subscribing to topic [webToCabinets2] with maximum QoS [0]. 
MQTT [ip:port] [cc432d1b33936e5c206f] Sent data over the socket: �webToCabinets2
MQTT [ip:port] [cc432d1b33936e5c206f] Sending disconnect package to the broker.
MQTT [ip:port] [cc432d1b33936e5c206f] Sent data over the socket: �
MQTT [ip:port] [cc432d1b33936e5c206f] Closing the socket to the broker.
```

**Comment by Namoshek** [comment]
The last five lines tell that you are not calling `$client->loop(true);`

**Comment by StefanCiobanu1989** [comment]
**"if '$client->loop(true);' was to be removed then the 500 error goes away but still no message come back from the broker. "**, i removed it otherwise after a couple of minutes i would get a 500 server error

**Comment by Namoshek** [comment]
You can't just remove code and expect it to still work. :wink:

PHP code by itself is entirely synchronous, making it somewhat difficult to implement a data receiving client. When we call `$client->subscribe($topic, $callback)`, all it does is sending a subscription request to the broker. This request is not acknowledged at this point in time (i.e. the broker has not yet confirmed it will send messages for this subscription). This is where `$client->loop(true)` comes in. It is responsible for retrieving and handling messages sent by the broker, for example `SUBACK` and `PUBLISH`.

If there is no exit condition somewhere in the code, `$client->loop(true)` will run forever (or until the program is killed). In a normal web application, where such an MQTT script is called via an HTTP request, this is an issue, since requests normally have a `max_execution_time`. And on top, most deployments have a reverse proxy in front which also adds a timeout, with the difference that it will not stop the script but only return a `504 Gateway Timeout` to the client.

So in short: do not use the MQTT client for subscriptions in a request-based application, until you know exactly what you are doing. The only use case for subscriptions is the RPC (_Remote Procedure Call_) pattern. You can find examples for this [here](https://github.com/php-mqtt/client/issues/66) and [here](https://github.com/php-mqtt/laravel-client/issues/13).
